compiler/hash-semantics/src/passes/discovery/visitor.rs

```diff
@@ -207,41 +211,114 @@ impl<E: SemanticEnv> ast::AstVisitor for DiscoveryPass<'_, E> {
+
+                // Lookup if there is a discriminant value for the given variant and
+                // then set it as the last discriminant. If
+                // the variant does not have an explicit
+                // discriminant, then we simply use the last known discriminant and
+                // increment it by one.
+                let (discriminant, origin) = if let Some(discr_annot) =
+                    attr_store().get_attr(variant.id(), attrs::DISCRIMINANT)
+                {
+                    let origin = NodeOrigin::Given(discr_annot.origin);
+                    let arg = discr_annot.get_arg(0).unwrap();
+
+                    let const_val = arg.as_int().value().value;
+                    let raw_val = const_val.as_u128();
+                    let mut discriminant = Discriminant {
+                        value: raw_val,
+                        ty: *discr_ty,
+                        kind: DiscriminantKind::Explicit,
+                    };
+
+                    if discriminant.has_overflowed(self.env) {
+                        return Err(SemanticError::EnumDiscriminantOverflow {
+                            location: origin.span().unwrap(),
+                            annotation_origin: discr_ty.span(),
+                            discriminant,
+                        });
+                    } else if discr_ty.is_signed() {
+                        // @@Hack @@TIRConsts: we manually truncate the value of the discriminant to
+                        // being a u128 so that we can detect whether it overflows
+                        // or not. However, we should remove this when we can use
+                        // the `Const` format in the TIR which will automatically perform
+                        // the truncation.
+                        discriminant.value =
+                            discr_ty.size(self.target().ptr_size()).sign_extend(raw_val);
+                    }
+
+                    prev_discr = Some(discriminant);
+                    (create_term_from_integer_lit(const_val, origin), origin)
+                } else {
+                    let origin = NodeOrigin::Given(variant.id());
+
+                    // We want to check if we are starting from the very beginning which
+                    // should actually use the zero discriminant.
+                    let (next_discr, _) = if prev_discr.is_some() {
+                        prev.checked_add(self.env, 1)
+                    } else {
+                        (prev, false)
+                    };
+
+                    // @@Hack: we convert the `u128` into an int constant
+                    // with the given discriminant type, later this should be
+                    // replaced by the new constant representation.
+                    let constant = IntConstant::from_scalar(
+                        next_discr.value,
+                        *discr_ty,
+                        self.target().ptr_size(),
```

Comment:
Could all this be factored out to a separate function? It seems a bit out-of-place here

Reply:
Yes it should indeed, I am just unsure of where to actually put the logic. Also, check the top comment I left, maybe there should be a better name for the file that contains all of the discriminant handling logic

Reply:
Maybe `utils.rs` should be renamed to `discriminants.rs` and this code can be moved in there too?